bug: compiler-crashing ICE bug on nested `const fn` (src/function/fetch.rs:178:21, scarb/src/ops/compile.rs:238:14)

# Bug Report
**[Note: I would be happy to help add a PR to fix this issue. Feel free to assign it to me and I will accordingly fix the problem]**

**Cairo version:**

<!-- Please specify commit or tag version. -->
scarb 2.18.0 (e6144df0f 2026-04-21)
cairo: 2.18.0 (https://crates.io/crates/cairo-lang-compiler/2.18.0)
sierra: 1.8.0
arch: aarch64-apple-darwin



**Current behavior:**

<!-- Describe how the bug manifests. -->


`scarb build` causes compiler ICE when a `const fn` is declared inside another `const fn`'s body. The parser does report its usual recovery diagnostics (E1001 / E1002 / E1004 / E1013/etc) for the malformed nesting, but compilation continues into the semantic layer, where `priv_free_function_body_data` enters a query cycle and panics.


```
   Compiling scratch v0.1.0 (bug_poc/Scarb.toml)
error[E1004]: Unexpected token, expected ':' followed by a type.
 --> bug_poc/src/lib.cairo:13:13
    const fn inner() -> felt252 { 7 }
            ^

error[E1001]: Missing token '='.
 --> bug_poc/src/lib.cairo:15:2
}
 ^

error[E1002]: Missing tokens. Expected an expression.
 --> bug_poc/src/lib.cairo:15:2
}
 ^

error[E1001]: Missing token ';'.
 --> bug_poc/src/lib.cairo:15:2
}
 ^

error[E1013]: 'fn' is a reserved identifier.
 --> bug_poc/src/lib.cairo:13:11
    const fn inner() -> felt252 { 7 }
          ^^

error[E1001]: Missing token '}'.
 --> bug_poc/src/lib.cairo:18:14
#[executable]
             ^

error[E1026]: Missing tokens. Expected a statement after attributes.
 --> bug_poc/src/lib.cairo:18:14
#[executable]
             ^


thread '<unnamed>' (3106047) panicked at /Users/runner/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/salsa-0.26.0/src/function/fetch.rs:178:21:
dependency graph cycle when querying priv_free_function_body_data(Id(a000)), set cycle_fn/cycle_initial to fixpoint iterate.
Query stack:
[
    module_semantic_diagnostics_tracked(Id(b800)),
    priv_free_function_body_data(Id(a000)),
    type_size_info_tracked(Id(1f000)),
    priv_type_is_fully_concrete_tracked(Id(1f000)),
    priv_module_semantic_data(Id(c818)),
    get_children(Id(2d2db)),
    module_sub_files(Id(15c2c)),
    get_children(Id(2d2ce)),
    file_syntax_data(Id(1582c)),
    new_root_node(Id(16c2c)),
    priv_raw_file_content(Id(1582c)),
    priv_raw_file_content(Id(15829)),
]
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

thread 'main' (3106016) panicked at scarb/src/ops/compile.rs:238:14:
Compiler thread has panicked.: Any { .. }
```

**Expected behavior:**

<!-- Describe what you expect the behavior to be without the bug. -->

normal  diagnostics (E1001 / E1002 / E1004 / E1013), no ICE error



**Steps to reproduce:**

<!-- Explain the steps required to duplicate the issue, especially if you are able to provide a sample application. -->

just run `scarb build` on the following PoC:

**Related code:**

<!-- If you are able to illustrate the bug or feature request with an example, please provide it here. -->

Scarb.toml

```
[package]
name = "scratch"
version = "0.1.0"
edition = "2025_12"

[[target.executable]]

[cairo]
enable-gas = false

[dependencies]
cairo_execute = "2.18.0"
```


lib.cairo:
```
const fn outer() -> felt252 {
    const fn inner() -> felt252 { 7 }
    inner()
}
const X: felt252 = outer();

#[executable]
fn main() -> felt252 { X }
```

**Other information:**

<!-- List any other information that is relevant to your issue. Related issues, suggestions on how to fix, Stack Overflow links, forum links, etc. -->


## Root Cause and Mitigation
Some preliminary investigation:
- `fn` nested in `fn` body  produces E1000 / E1001 / E1012 ,  **clean errors, no panic** 
- `const fn` nested in `fn` body produces E1001 / E1002 / E1004 / E1013,  **clean errors, no panic** 
- `const fn` nested in `#[executable]` `main`  produce E1001 / E1002 / E1004 / E1013, **clean errors, no panic** 
BUT, in our case considered in this issue report, `const fn` nested in `const fn` body produced E1001 / E1002 / E1004 / E1013, **PLUS ICE panic**.


> [**Note: the following suggestions are from an LLM when I was trying to double check to confirm the bug; I was hoping this could help making your debugging easier, but I haven't fully checked this part yet (looks good to me overall based on my sanity check) so this suggestion might not necessarily be reliable**]


**Defensive fix at salsa boundary (recommended ship fix).** Add `#[salsa::cycle_fn(... cycle_initial=...)]` to `priv_free_function_body_data`. The cycle-initial returns empty body-data + an `InternalCompilerError` diagnostic instead of panicking. Since the AST is already malformed (parser emitted multiple error codes), surfacing one more diagnostic is fine. ~20 LoC.

--- a/crates/cairo-lang-semantic/src/expr/test_data/constant
+++ b/crates/cairo-lang-semantic/src/expr/test_data/constant
@@ -967,3 +967,61 @@ error[E2127]: This expression is not supported as constant.
  --> lib.cairo:5:26
     const VALUE: bool = !F::VALUE;
                          ^^^^^^^^
+
+//! > ==========================================================================
+
+//! > Test nested const fn inside const fn emits parser diagnostics without ICE.
+
+//! > test_runner_name
+test_function_diagnostics(expect_diagnostics: true)
+
+//! > module_code
+const fn outer() -> felt252 {
+    const fn inner() -> felt252 { 7 }
+    inner()
+}
+const X: felt252 = outer();
+
+//! > function_code
+fn foo() -> felt252 {
+    X
+}
+
+//! > function_name
+foo
+
+//! > expected_diagnostics
+error[E1004]: Unexpected token, expected ':' followed by a type.
+ --> lib.cairo:2:13
+    const fn inner() -> felt252 { 7 }
+            ^
+
+error[E1001]: Missing token '='.
+ --> lib.cairo:4:2
+}
+ ^
+
+error[E1002]: Missing tokens. Expected an expression.
+ --> lib.cairo:4:2
+}
+ ^
+
+error[E1001]: Missing token ';'.
+ --> lib.cairo:4:2
+}
+ ^
+
+error[E1013]: 'fn' is a reserved identifier.
+ --> lib.cairo:2:11
+    const fn inner() -> felt252 { 7 }
+          ^^
+
+error[E1001]: Missing token '}'.
+ --> lib.cairo:5:28
+const X: felt252 = outer();
+                           ^
+
+error[E0006]: Identifier not found.
+ --> lib.cairo:7:5
+    X
+    ^

--- a/crates/cairo-lang-semantic/src/items/constant.rs
+++ b/crates/cairo-lang-semantic/src/items/constant.rs
@@ -885,7 +885,9 @@ impl<'a, 'r, 'mt> ConstantEvaluateContext<'a, 'r, 'mt> {
             return calc_result;
         }
 
-        let imp = extract_matches!(concrete_function.generic_function, GenericFunctionId::Impl);
+        let GenericFunctionId::Impl(imp) = concrete_function.generic_function else {
+            return to_missing(skip_diagnostic());
+        };
         let bool_value = |condition: bool| {
             if condition { self.true_const } else { self.false_const }
         };
@@ -938,12 +940,7 @@ impl<'a, 'r, 'mt> ConstantEvaluateContext<'a, 'r, 'mt> {
                 )
                 .intern(db);
             }
-            _ => {
-                unreachable!(
-                    "Unexpected function call in constant lowering: `{:?}`",
-                    expr.function.debug(db)
-                )
-            }
+            _ => return to_missing(skip_diagnostic()),
         };
         if expr.ty == self.felt252 {
             ConstValue::Int(canonical_felt252(&value), expr.ty).intern(db)

--- a/crates/cairo-lang-semantic/src/items/free_function.rs
+++ b/crates/cairo-lang-semantic/src/items/free_function.rs
@@ -4,7 +4,7 @@ use cairo_lang_defs::db::DefsGroup;
 use cairo_lang_defs::ids::{
     FreeFunctionId, FunctionTitleId, LanguageElementId, LookupItemId, ModuleItemId,
 };
-use cairo_lang_diagnostics::{Diagnostics, Maybe, MaybeAsRef};
+use cairo_lang_diagnostics::{Diagnostics, Maybe, MaybeAsRef, skip_diagnostic};
 use cairo_lang_syntax::attribute::structured::{Attribute, AttributeListStructurize};
 use cairo_lang_syntax::node::{TypedStablePtr, TypedSyntaxNode};
 use cairo_lang_utils::Intern;
@@ -124,7 +124,7 @@ fn free_function_declaration_data<'db>(
 }
 
 /// Query implementation of [FreeFunctionSemantic::priv_free_function_body_data].
-#[salsa::tracked(returns(ref))]
+#[salsa::tracked(returns(ref), cycle_fn=priv_free_function_body_data_cycle, cycle_initial=priv_free_function_body_data_initial)]
 fn priv_free_function_body_data<'db>(
     db: &'db dyn Database,
     free_function_id: FreeFunctionId<'db>,
@@ -174,6 +174,26 @@ fn priv_free_function_body_data<'db>(
         resolver_data,
         body: FunctionBody { arenas, body_expr },
     })
+}
+
+/// Cycle handling for [FreeFunctionSemantic::priv_free_function_body_data].
+fn priv_free_function_body_data_cycle<'db>(
+    _db: &'db dyn Database,
+    _cycle: &salsa::Cycle<'_>,
+    last_provisional_value: &Maybe<FunctionBodyData<'db>>,
+    _value: Maybe<FunctionBodyData<'db>>,
+    _free_function_id: FreeFunctionId<'db>,
+) -> Maybe<FunctionBodyData<'db>> {
+    last_provisional_value.clone()
+}
+
+/// Cycle handling for [FreeFunctionSemantic::priv_free_function_body_data].
+fn priv_free_function_body_data_initial<'db>(
+    _db: &'db dyn Database,
+    _id: salsa::Id,
+    _free_function_id: FreeFunctionId<'db>,
+) -> Maybe<FunctionBodyData<'db>> {
+    Err(skip_diagnostic())
 }
 
 /// Trait for free function-related semantic queries.

